source/core_json.c

```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
         }
     }
 
-    if( ret == JSONSuccess )
+    if( ret == true )
     {
-        /* String values and collections include their surrounding
-         * demarcation.  If the value is a string, strip the quotes. */
-        if( buf[ value ] == '"' )
+        *outValue = &buf[ value ];
+        *outValueLength = valueLength;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Find an index in a JSON array and output a pointer to its value.
+ *
+ * @param[in] buf  The buffer to search.
+ * @param[in] max  size of the buffer.
+ * @param[in] queryIndex  The index to search for.
+ * @param[out] outValue  A pointer to receive the address of the value found.
+ * @param[out] outValueLength  A pointer to receive the length of the value found.
+ *
+ * Iterate over the values of an array, looking for a matching index.
+ *
+ * @return true if the queryIndex is found and the value output;
+ * false otherwise.
+ *
+ * @note Parsing stops upon finding a match.
+ */
+static bool_ arraySearch( char * buf,
+                          size_t max,
+                          uint32_t queryIndex,
+                          char ** outValue,
+                          size_t * outValueLength )
+{
+    bool_ ret = false;
+    size_t i = 0, value, valueLength;
+    uint32_t j = 0;
+
+    assert( buf != NULL );
+    assert( ( outValue != NULL ) && ( outValueLength != NULL ) );
+
+    skipSpace( buf, &i, max );
+
+    if( ( i < max ) && ( buf[ i ] == '[' ) )
+    {
+        i++;
+        skipSpace( buf, &i, max );
+
+        while( i < max )
         {
-            value++;
-            valueLength -= 2U;
+            if( nextValue( buf, &i, max, &value, &valueLength ) != true )
+            {
+                break;
+            }
+
+            if( j == queryIndex )
+            {
+                ret = true;
+                break;
+            }
+
+            if( skipSpaceAndComma( buf, &i, max ) != true )
+            {
+                break;
+            }
+
+            j++;
         }
+    }
 
+    if( ret == true )
+    {
         *outValue = &buf[ value ];
         *outValueLength = valueLength;
     }
-    else if( ret == JSONPartial )
+
+    return ret;
+}
+
+/**
+ * @brief Advance buffer index beyond a query key subpart.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ *
+ * @return true if a valid string was present;
+ * false otherwise.
+ */
+#define JSON_QUERY_KEY_SEPARATOR    '.'
+#define isSeparator_( x )    ( ( x ) == JSON_QUERY_KEY_SEPARATOR )
+static bool_ skipQueryKey( const char * buf,
```

Comment:
`advanceToQueryKeySubpart` is an easier name to understand the purpose of this function (even though it is wordy). **skipQueryKey** is ambiguous (imo).
```suggestion
static bool_ advanceToQueryKeySubpart( const char * buf,
```

Reply:
changed to `skipQueryPart`